### PR TITLE
Add helm dependency update

### PR DESCRIPTION
**Pull request by WissemChb**
##### SUMMARY

Execute the helm dependency update under the hood when found `dependencies` block in Chart.yaml file.
Support the execution of:

- Standalone dependency update by executing: `helm dependency update CHART` 
- Inline dependency update when specifying the helm `chart_repo_url` by adding `--dependency-update` to the `helm install` command.

##### ISSUE TYPE

- Feature Pull Request #191 

##### COMPONENT NAME

helm, helm_template
##### ADDITIONAL INFORMATION

- There is a doc generated for `history_max` option for the helm module. I think that is not generated in the previous PR #164.
- There is others changes affect the `docs/` folder when I run the `collection_prep_add_docs -p . ` command. These changes are added in the last commit  64eab40. I let you decide rather we keep the commit or remove it.

- The `--dependency-update` insertion option is tested used a local helm chart repository create via docker. So here are the tasks that test this feature.  Maybe if we create a GitHub repository for the helm chart, we can add this test code in the CI pipeline.

```paste below
# Test The update dependency with chart_repo_url
- name: "Test chart without dependencies block and chart_repo_url defined"
  block:
    - name: "Test chart without dependencies block and chart_repo_url defined"
      helm:
        binary_path: "{{ helm_binary }}"
        name: test
        chart_ref: "ingress-nginx"
        chart_repo_url: https://kubernetes.github.io/ingress-nginx
        chart_version: "{{ chart_source_version | default(omit) }}"
        namespace: "{{ helm_namespace }}"
        create_namespace: yes
      register: release

    - assert:
        that:
          - "'--dependency-update' not in release.command"
          - "'upgrade' in release.command"
        success_msg: "Command does not contains '--dependency-update' options"
        fail_msg: "Command contains '--dependency-update' options"

- name: "Test chart with dependencies block and chart_repo_url defined and replace True"
  block:
    - name: "Test chart with dependencies block and chart_repo_url defined and replace True"
      helm:
        binary_path: "{{ helm_binary }}"
        name: test1
        chart_ref: "dep_up"
        chart_repo_url: http://repo:8080/charts
        chart_version: "{{ chart_source_version | default(omit) }}"
        namespace: "{{ helm_namespace }}"
        create_namespace: yes
        replace: true
      register: release
    - debug: var=release
    - assert:
        that:
          - "'--dependency-update' in release.command"
          - "'install' in release.command"
        success_msg: "Command contains '--dependency-update' options with helm install command"
        fail_msg: "Command not contains '--dependency-update' with helm install command"

- name: "Test chart with dependencies block and chart_repo_url defined and replace False fails"
  block:
    - name: "Test chart with dependencies block and chart_repo_url defined and replace False fails"
      helm:
        binary_path: "{{ helm_binary }}"
        name: test2
        chart_ref: "dep_up"
        chart_repo_url: http://repo:8080/charts
        chart_version: "{{ chart_source_version | default(omit) }}"
        namespace: "{{ helm_namespace }}"
        create_namespace: yes
        replace: false
      register: release
      ignore_errors: true

    - assert:
        that:
          - release.failed
          - release.msg == "'--dependency-update' hasn't been supported yet with 'helm upgrade'. Please use 'helm install' instead by adding 'replace' option"
        success_msg: "Command build fail when adding  '--dependency-update' with the helm upgrade command"
```

### CODE REVIEW

**Comment by gravesm** [comment]
@WissemChb thanks for the PR! This is looking like a good start. I think the better way to do this is to add a `dependency_update` parameter as you initially suggested in your issue, though. This PR forces the update which would change the existing behavior and also not allow any control over whether or not to do the update.

With regards to the docs changes, you should restore all the deleted .rst files. The doc generation command is pretty rough and ideally shouldn't be deleting those. The other changes you can keep. We always generate the docs right before a release but it's fine to go ahead and run it now, too.

**Comment by WissemChb** [comment]
> @WissemChb thanks for the PR! This is looking like a good start. I think the better way to do this is to add a `dependency_update` parameter as you initially suggested in your issue, though. This PR forces the update which would change the existing behavior and also not allow any control over whether or not to do the update.
> 
> With regards to the docs changes, you should restore all the deleted .rst files. The doc generation command is pretty rough and ideally shouldn't be deleting those. The other changes you can keep. We always generate the docs right before a release but it's fine to go ahead and run it now, too.

Hello @gravesm, Thank you for your reply. I'll take into consideration your suggestions. Initially, I considered adding the `dependency_update` options but when I go through the dev process I decide to simplify the module params and options. But you are right it's better to add more control over the update process.

File: molecule/default/roles/helm/tasks/test_up_dep.yml [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+      register: stat_result
+
+    - assert:
+        that:
+          - stat_result.stat.exists
+        success_msg: "subchart exist in the chart directory"
+        fail_msg: "subchart not exist in the charts directory"
+
+# Clean
+- name: "Clean"
```
Comment (gravesm) [inline molecule/default/roles/helm/tasks/test_up_dep.yml:79]: This needs to be in an `always` block. The best way to do this will be to put everything in this file except this block into a single block and then put this part into an `always` block at the bottom. Look at tests_helm_diff.yml for an example.

File: molecule/default/roles/helm/tasks/test_up_dep.yml [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+    - name: Test Helm subchart not pulled when dependency_update false
+      helm_template:
+        binary_path: "{{ helm_binary }}"
+        chart_ref: "/tmp/test-chart"
+        chart_version: "{{ chart_source_version | default(omit) }}"
+        dependency_update: false
+        output_dir: "/tmp"
+      register: result
+
+    - name: "Check if the subchart exist in chart"
```
Comment (gravesm) [inline molecule/default/roles/helm/tasks/test_up_dep.yml:162]: This should be testing that `/tmp/test-chart/charts` does not exist.

File: plugins/modules/helm.py [hunk 3]
```diff
@@ -46,6 +46,16 @@
...
+  dependency_update:
+    description:
+      - Run standelone C(helm dependency update CHART) before the operation.
+      - Run inline C(--dependency-update) with C(helm install) command. This feature is not supported yet with the C(helm upgrade) command.
+      - So we should consider to use I(dependency_update) options with I(replace) option enabled when specifying I(chart_repo_url).
+      - The I(dependency_update) option require the add of C(dependencies) block in C(Chart.yaml/requirements.yaml) file.
+      - For more information please visit U(https://helm.sh/docs/helm/helm_dependency/)
+    default: false
+    type: bool
+    aliases: [ dep_up ]
```
Comment (gravesm) [inline plugins/modules/helm.py:58]:
```suggestion
    aliases: [ dep_up ]
    version_added: "2.2.0"
```

File: plugins/modules/helm_template.py [hunk 4]
```diff
@@ -45,6 +45,14 @@
...
     required: false
     type: str
+  dependency_update:
+    description:
+      - Run helm dependency update before the operation.
+      - The I(dependency_update) option require the add of C(dependencies) block in C(Chart.yaml/requirements.yaml) file.
+      - For more information please visit U(https://helm.sh/docs/helm/helm_dependency/)
+    default: false
+    type: bool
+    aliases: [ dep_up ]
```
Comment (gravesm) [inline plugins/modules/helm_template.py:55]:
```suggestion
    aliases: [ dep_up ]
    version_added: "2.2.0"
```

**Review by gravesm — CHANGES_REQUESTED** [review]
Thanks for your work on this, and for adding tests! In addition to inline comments we'll need a changelog fragment: https://docs.ansible.com/ansible/latest/community/development_process.html#changelogs

File: plugins/modules/helm.py [hunk 5]
```diff
@@ -612,17 +634,38 @@ def main():
...
 
         # Fetch chart info to have real version and real name for chart_ref from archive, folder or url
         chart_info = fetch_chart_info(module, helm_cmd, chart_ref)
 
+        if dependency_update:
+            if not chart_info.get('dependencies'):
+                msg_fail = (f"No subchart will be pulled for {chart_ref}. "
+                            "Please make sure to add dependencies block in Chart.yaml or requirements.yaml. "
+                            "For more information please visite https://helm.sh/docs/helm/helm_dependency/")
+                module.fail_json(msg=msg_fail)
```
[the remark this answers is not in the document]
Comment (WissemChb) [reply plugins/modules/helm.py:649]: I hesitated between adding only a warning message or fail the execution. And I think I chose the wrong option. So I'll add a warning message instead.

[the remark this answers is not in the document]
Comment (WissemChb) [reply plugins/modules/helm.py:649]: I hesitated between a warn message or a fail message and I think that I chose the wrong option. So l'il replace this block with a warn message.

File: plugins/modules/helm.py [hunk 6]
```diff
@@ -612,17 +634,38 @@ def main():
...
+            # we can get rid of this method and use only '--dependency-update'
+            # option. Please see https://github.com/helm/helm/pull/8810
+            if not chart_repo_url:
+                run_dep_update(module, helm_cmd_common, chart_ref)
+            else:
+                module.warn("This is a not stable feature with 'chart_repo_url'. Please consider to use dependency update with on-disk charts")
+                if not replace:
+                    msg_fail = ("'--dependency-update' hasn't been supported yet with 'helm upgrade'. "
+                                "Please use 'helm install' instead by adding 'replace' option")
+                    module.fail_json(msg=msg_fail)
```
[the remark this answers is not in the document]
Comment (WissemChb) [reply plugins/modules/helm.py:663]: This part is becoming very confusing for me. for 2 reasons:
1. What I found in Ansible doc is that if I make `dependency_update` and `replace` mutually exclusive I can't enable both of them and that is not the case because when the `chart_ref_url` is defined or `chart_ref` is an URL I need to enable `replace` option when `dependency_update` is enabled to use `helm install` instead of `helm update`(not support `--dependency-update` option. I'm sorry, this is my first ansible module creation maybe I need more details about your suggestion.
2. After doing a deep search I realised that we haven't the need to do `dependency update` with remote charts because to package them we need to pull the subchart by running `helm package --dependency-update dep-up` So, if I'm not mistaken, all remote package that depends on other charts have their subchart pulled already within them.

File: plugins/modules/helm.py [hunk 7]
```diff
@@ -612,17 +634,38 @@ def main():
...
+                            "Please make sure to add dependencies block in Chart.yaml or requirements.yaml. "
+                            "For more information please visite https://helm.sh/docs/helm/helm_dependency/")
+                module.fail_json(msg=msg_fail)
+
+            # Can't use '--dependency-update' with 'helm upgrade' that is the
+            # default chart install method, so if chart_repo_url is defined
+            # we can't use the dependency update command. But, in the near future
+            # we can get rid of this method and use only '--dependency-update'
+            # option. Please see https://github.com/helm/helm/pull/8810
+            if not chart_repo_url:
```
[the remark this answers is not in the document]
Comment (WissemChb) [reply plugins/modules/helm.py:656]: related to the previous reply

File: docs/kubernetes.core.k8s_json_patch_module.rst [hunk 8]
```diff
@@ -18,7 +18,7 @@ Version added: 2.0.0
 Synopsis
 --------
 - This module is used to apply RFC 6902 JSON patch operations only.
-- Use the :ref:`k8s <k8s_module>` module for strategic merge or JSON merge operations.
+- Use the :ref:`kubernetes.core.k8s <kubernetes.core.k8s_module>` module for strategic merge or JSON merge operations.
```
Comment (abikouo) [inline docs/kubernetes.core.k8s_json_patch_module.rst:21]: This has nothing to do with this PR, could you please revert this file ? 
Thanks

```suggestion
- Use the :ref:`k8s <k8s_module>` module for strategic merge or JSON merge operations.
```

File: docs/kubernetes.core.k8s_log_module.rst [hunk 9]
```diff
@@ -383,6 +383,22 @@ Parameters
...
+                        <span style="color: purple">string</span>
+                    </div>
+                    <div style="font-style: italic; font-size: small; color: darkgreen">added in 2.2.0</div>
+                </td>
+                <td>
+                </td>
+                <td>
+                        <div>A relative time in seconds before the current time from which to show logs.</div>
+                </td>
+            </tr>
```
Comment (abikouo) [inline docs/kubernetes.core.k8s_log_module.rst:401]: please revert this modification. the doc for this module should be modified

File: docs/kubernetes.core.k8s_lookup.rst [hunk 10]
```diff
@@ -381,6 +381,12 @@ Parameters
     <br/>
 
 
+Notes
+-----
+
+.. note::
+   - While querying, please use ``query`` or ``lookup`` format with ``wantlist=True`` to provide an easier and more consistent interface. For more details, see https://docs.ansible.com/ansible/latest/plugins/lookup.html#forcing-lookups-to-return-lists-query-and-wantlist-true.
+
```
Comment (abikouo) [inline docs/kubernetes.core.k8s_lookup.rst:389]: same for this file, should not be modified

File: plugins/modules/helm.py [hunk 11]
```diff
@@ -46,6 +46,17 @@
       - Chart version to install. If this is not specified, the latest version is installed.
     required: false
     type: str
+  dependency_update:
+    description:
+      - Run standelone C(helm dependency update CHART) before the operation.
```
Comment (abikouo) [inline plugins/modules/helm.py:51]:
```suggestion
      - Run standalone C(helm dependency update CHART) before the operation.
```

File: plugins/modules/helm.py [hunk 12]
```diff
@@ -612,17 +635,39 @@ def main():
         if chart_version is not None:
             helm_cmd += " --version=" + chart_version
 
-        if chart_repo_url is not None:
+        if chart_repo_url:
```
Comment (abikouo) [inline plugins/modules/helm.py:638]: why this change ? 
In some case it is better to check ``None`` value instead of implicit conversion of str to bool

File: plugins/modules/helm.py [hunk 13]
```diff
@@ -612,17 +635,39 @@ def main():
...
+                            "Please consider add dependencies block or disable dependency_update to remove this warning.")
+
+            # Can't use '--dependency-update' with 'helm upgrade' that is the
+            # default chart install method, so if chart_repo_url is defined
+            # we can't use the dependency update command. But, in the near future
+            # we can get rid of this method and use only '--dependency-update'
+            # option. Please see https://github.com/helm/helm/pull/8810
+            if not chart_repo_url:
+                run_dep_update(module, helm_cmd_common, chart_ref)
+
```
Comment (abikouo) [inline plugins/modules/helm.py:656]:
```suggestion
      else:
            # Can't use '--dependency-update' with 'helm upgrade' that is the
            # default chart install method, so if chart_repo_url is defined
            # we can't use the dependency update command. But, in the near future
            # we can get rid of this method and use only '--dependency-update'
            # option. Please see https://github.com/helm/helm/pull/8810
            if not chart_repo_url:
                run_dep_update(module, helm_cmd_common, chart_ref)

```

File: plugins/modules/helm.py [hunk 14]
```diff
@@ -612,17 +635,39 @@ def main():
...
             helm_cmd += " --repo=" + chart_repo_url
 
         # Fetch chart info to have real version and real name for chart_ref from archive, folder or url
         chart_info = fetch_chart_info(module, helm_cmd, chart_ref)
 
+        if dependency_update:
+            if not chart_info.get('dependencies'):
+                module.warn("There is no dependencies block defined in Chart.yaml. Dependency update will not be performed. "
+                            "Please consider add dependencies block or disable dependency_update to remove this warning.")
+
```
Comment (abikouo) [inline plugins/modules/helm.py:648]: why not putting else here to avoid running ``helm dependency update`` ?

File: plugins/modules/helm.py [hunk 15]
```diff
@@ -612,17 +635,39 @@ def main():
...
+            # Can't use '--dependency-update' with 'helm upgrade' that is the
+            # default chart install method, so if chart_repo_url is defined
+            # we can't use the dependency update command. But, in the near future
+            # we can get rid of this method and use only '--dependency-update'
+            # option. Please see https://github.com/helm/helm/pull/8810
+            if not chart_repo_url:
+                run_dep_update(module, helm_cmd_common, chart_ref)
```
Comment (abikouo) [inline plugins/modules/helm.py:655]: please consider returning at least the result code ``rc`` which will validate that the dependency update has succeed or return error message if not

File: plugins/modules/helm.py [hunk 16]
```diff
@@ -352,6 +363,14 @@ def run_repo_update(module, command):
     rc, out, err = run_helm(module, repo_update_command)
 
 
+def run_dep_update(module, command, chart_ref):
+    """
+    Run dependency update
+    """
+    dep_update = command + " dependency update " + chart_ref
```
Comment (abikouo) [inline plugins/modules/helm.py:370]: this will fail when the ``chart_ref`` contains an URL, you need to check somewhere that it is not the case

File: plugins/modules/helm.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (abikouo) [reply plugins/modules/helm.py:654]: agree, ``chart_ref`` should be validated as non URL

**Comment by WissemChb** [comment]
@gravesm and @abikouo, I added what you suggested. Can you check, please ? Thank you very much for your the time to make the review.
> one of the three sanity tests fails in the Zuul pipeline can you please check and tell me what I missed because the test worked fine in local for me.

**Comment by abikouo** [comment]
recheck

**Review by abikouo — CHANGES_REQUESTED** [review]

**Comment by joestump** [comment]
Got bit by this today. Anything I can do to help?

**Comment by gravesm** [comment]
Sorry, this one appears to have fallen through the cracks. I've gone ahead and rebased, and fixed a few things to account for recent changes to our CI.

**Comment by gravesm** [comment]
recheck

**Review by abikouo — APPROVED** [review]
LGTM!!

**Review by gravesm — APPROVED** [review]
